src/Dom/Document.php

```diff
@@ -1397,14 +1602,49 @@ private function convertAmpEmojiAttribute($source)
             return $source;
         }
 
-        $this->usedAmpEmoji = $matches[2];
+        $htmlTag = $matches[0];
 
-        return preg_replace(
-            self::AMP_EMOJI_ATTRIBUTE_PATTERN,
-            '\1' . self::EMOJI_AMP_ATTRIBUTE_PLACEHOLDER . '="\3"',
-            $source,
-            1
+        // Extract attributes.
+        if (!preg_match('#^(<html)(\s[^>]+)>$#i', $htmlTag, $matches)) {
+            return $source;
+        }
+
+        // Split into individual attributes.
+        $attributes = array_map(
+            'trim',
+            array_filter(
+                preg_split(
+                    '#(\s+[^"\'\s=]+(?:=(?:"[^"]+"|\'[^\']+\'|[^"\'\s]+))?)#',
+                    $matches[2],
+                    -1,
+                    PREG_SPLIT_DELIM_CAPTURE
+                )
+            )
         );
+
+        foreach ($attributes as $index => $attribute) {
+            $attributeMatches = [];
+            if (
+                preg_match(
+                    '/^(' . Attribute::AMP_EMOJI_ALT . '|' . Attribute::AMP_EMOJI . ')(4(?:ads|email))?$/i',
+                    $attribute,
+                    $attributeMatches
+                )
+            ) {
+                $this->usedAmpEmoji = $attributeMatches[1];
+                $variant            = ! empty($attributeMatches[2]) ? $attributeMatches[2] : '';
+                $attributes[$index] = self::EMOJI_AMP_ATTRIBUTE_PLACEHOLDER . "=\"{$variant}\"";
+
+                $source = preg_replace(
+                    self::AMP_EMOJI_ATTRIBUTE_PATTERN,
+                    '<html ' . implode(' ', $attributes) . '>',
+                    $source
+                );
+                break;
+            }
+        }
+
+        return $source;
```

Comment:
This could be made more efficient by adapting into a `preg_replace_callback` where the following code is put inside the callback. In that way, there would not be a need to do an additional `preg_replace()` on the `$source`. It could be combined with teh `preg_match()` call above as well. Something like this I believe:

```php
    private function convertAmpEmojiAttribute($source)
    {
        $this->usedAmpEmoji = '';

        return preg_replace_callback(
            self::AMP_EMOJI_ATTRIBUTE_PATTERN,
            static function ( $matches ) {
                // Split into individual attributes.
                $attributes = array_map(
                    'trim',
                    array_filter(
                        preg_split(
                            '#(\s+[^"\'\s=]+(?:=(?:"[^"]+"|\'[^\']+\'|[^"\'\s]+))?)#',
                            $matches[2],
                            -1,
                            PREG_SPLIT_DELIM_CAPTURE
                        )
                    )
                );

                foreach ($attributes as $index => $attribute) {
                    $attributeMatches = [];
                    if (
                        preg_match(
                            '/^(' . Attribute::AMP_EMOJI_ALT . '|' . Attribute::AMP_EMOJI . ')(4(?:ads|email))?$/i',
                            $attribute,
                            $attributeMatches
                        )
                    ) {
                        $this->usedAmpEmoji = $attributeMatches[1];
                        $variant            = ! empty($attributeMatches[2]) ? $attributeMatches[2] : '';
                        $attributes[$index] = self::EMOJI_AMP_ATTRIBUTE_PLACEHOLDER . "=\"{$variant}\"";
                        break;
                    }
                }

                return '<html ' . implode(' ', $attributes) . '>';
            },
            $source,
            1
        );
    }
```

Note also the added limit of 1.

Reply:
I adapted to code to use a single `preg_replace_callback()`. Your code was already pretty close, but it needed a bit more fiddling with the regex pattern(s).

Just to be sure, I profiled the difference across the unit tests we have that cover emojis, and this is the result:

<img width="297" alt="Comparison from Untitled to Untitled - Blackfire - Google Chrome 2021-03-15 at 11 22 19 AM" src="https://user-images.githubusercontent.com/83631/111146532-754a2180-8581-11eb-9203-e36b562bf268.png">

So it's indeed a clear improvement overall.